libyul/backends/evm/SSAControlFlowGraphBuilder.cpp

```diff
@@ -535,7 +523,23 @@ void SSAControlFlowGraphBuilder::operator()(Leave const& _leaveStatement)
 void SSAControlFlowGraphBuilder::operator()(Block const& _block)
 {
 	ScopedSaveAndRestore saveScope(m_scope, m_info.scopes.at(&_block).get());
-	for (auto const& statement: _block.statements)
+	auto const filterFunctionDefinitions = [](bool _yieldDefinitionsOnly)
+	{
+		return [b=_yieldDefinitionsOnly](Statement const& _statement) -> bool { return !b ^ std::holds_alternative<FunctionDefinition>(_statement);};
+	};
+	// gather all function definitions so that they are visible to each other's subgraphs
+	for (auto const& statement: _block.statements | ranges::views::filter(filterFunctionDefinitions(true)))
+		std::visit(*this, statement);
+	// build the subgraphs
+	for (auto const& statement: _block.statements | ranges::views::filter(filterFunctionDefinitions(true)))
+	{
+		auto const* functionDefinition = std::get_if<FunctionDefinition>(&statement);
+		yulAssert(functionDefinition);
+		Scope::Function const& function = lookupFunction(functionDefinition->name);
+		buildFunctionGraph(&function, functionDefinition);
+	}
```

Comment:
Any reason for not doing *this* in the ``operator()(FunctionDefinition)`` visitor instead? That'd avoid visiting everything *three* times.
You'd just need to move what's currently in that visitor into a separate ``registerFunctionDefinition`` (or the like) function and call that instead of ``std::visit(*this`` in the initial iteration.

Reply:
Absolutely right!